vcs/git/gogit/repo.go

```diff
@@ -0,0 +1,354 @@
+package git
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/shazow/go-git"
+	"github.com/shazow/go-vcs/vcs/gitcmd"
+	"golang.org/x/tools/godoc/vfs"
+	"sourcegraph.com/sourcegraph/go-vcs/vcs"
+	"sourcegraph.com/sqs/pbtypes"
+)
+
+func init() {
+	// Overwrite the git opener to return repositories that use the
+	// gogits native-go implementation.
+	vcs.RegisterOpener("git", func(dir string) (vcs.Repository, error) {
+		return Open(dir)
+	})
+}
+
+// Repository is a git VCS repository.
+type Repository struct {
+	// TODO: Do we need locking?
+	repo *git.Repository
+	// fallback is used for features that are not implemented here yet.
+	fallback *gitcmd.Repository
+}
+
+func (r *Repository) RepoDir() string {
+	return r.repo.Path
+}
+
+func (r *Repository) GitRootDir() string {
+	return r.RepoDir()
+}
+
+func (r *Repository) String() string {
+	return fmt.Sprintf("git (gogit) repo at %s", r.RepoDir())
+}
+
+func Open(dir string) (*Repository, error) {
+	if _, err := os.Stat(filepath.Join(dir, ".git")); !os.IsNotExist(err) {
+		// Append .git to path
+		dir = filepath.Join(dir, ".git")
+	}
+
+	repo, err := git.OpenRepository(dir)
+	if err != nil {
+		// FIXME: Wrap in vcs error?
```

Comment:
Fixed in 22729eab8b2e4c696fbcfd12d51ea4472e7a696a, added an extra message into the "op" string to avoid swallowing the underlying error (otherwise hard to debug if a specific portion is corrupted and we swallow the underlying error).